### PR TITLE
Transfering code to containers

### PR DESCRIPTION
**Pull request by nhorlava**
Key features:

1. The container recipe with tf 2.14 was created to run SynthSeg (see “container/README.md”). Examples of running slurm jobs with a container are in the “scripts/slurm/container_example”.

2. Code was modified to be reproducible via setting seed before starting the training: https://github.com/halirutan/SynthSeg/blob/3275e20cfec011a4656e127b46e93133ce7dc123/SynthSeg/training_with_tfrecords.py#L28-L31

3. The option to resume training from scratch was implemented:
	- the model checkpoint is now saved in the ".keras" format instead of ".h5" format in order to ensure correct loading of optimizer states (see for more details: https://www.tensorflow.org/tutorials/keras/save_and_load#new_high-level_keras_format)
	- To ensure that we resume training from the same batch and not from the beginning of the dataset, we access the number of batches that model has already seen from the loaded model checkpoint https://github.com/halirutan/SynthSeg/blob/3275e20cfec011a4656e127b46e93133ce7dc123/SynthSeg/training_with_tfrecords.py#L243
	 and skip this number of batched prior to the start of the training:https://github.com/halirutan/SynthSeg/blob/3275e20cfec011a4656e127b46e93133ce7dc123/SynthSeg/training_with_tfrecords.py#L174
	 This is done because `steps_per_epochs` in the fit loop is provided, thus one epoch runs not until the input dataset is exhausted, but for a specified `N` number of steps. The next epoch then starts not from the beginning of the dataset, but from the next batch `N+1`. Therefore, to ensure that when resuming the training we start not from the beginning of the dataset but from the correct `N+1` batch, we skip the first `N` batches. 
4. Additionally, custom callback was implemeted for profiling time spent for batch/epoch/fit loop: https://github.com/halirutan/SynthSeg/blob/3275e20cfec011a4656e127b46e93133ce7dc123/SynthSeg/training_with_tfrecords.py#L254
 That was done to ensure that the model training scales when using the distributed training and multiple GPUs when running the code in a container.

### CODE REVIEW

File: SynthSeg/training_options.py [hunk 1]
```diff
@@ -368,6 +368,24 @@ class TrainingOptions(Serializable, OptionsBase):
...
     """
 
+    scaling_type: str = "weak"
+    """
+    Possible values: "weak", "strong". 
+    Option that was used when profiling the distributed training. 
+    When set to "strong" will set batch_size to the batch_size/num_gpu (to see same amount of data as withg one gpu). 
+    When set to "weak" batch size will stay the same.
+    """
+
```
Comment (dcfidalgo) [inline SynthSeg/training_options.py]: I would remove this option, we do not really want to be able to "emulate" both cases, no study needed for this project.

File: SynthSeg/training_with_tfrecords.py [hunk 2]
```diff
@@ -183,21 +228,87 @@ def load_model(
...
             model.load_weights(checkpoint, by_name=True)
-
-    return model, is_compiled, init_epoch
+            print("Loading weights only")
+
+    return model, is_compiled, init_epoch, init_batch_idx
+
+
+
+class TimingCallback(tf.keras.callbacks.Callback):
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: I don't think we need this callback, I would prefer a more lightweight PR where we address the two main issues: container and resuming

File: SynthSeg/training_with_tfrecords.py [hunk 3]
```diff
@@ -22,6 +25,10 @@ class NullStrategy:
     def scope():
         return nullcontext()
 
+def set_seed(seed = 42):
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: Maybe we could expose the `seed` argument in the training options.

File: SynthSeg/training_with_tfrecords.py [hunk 4]
```diff
@@ -165,14 +199,25 @@ def load_model(
...
+                                                for el in glob.glob(f"{checkpoint}/*.keras", recursive=True)])
+            checkpoint = Path(max(list(files[files.ckpt==files.ckpt.max()].fullpath), key=os.path.getctime)) 
+            print(f"Model will continue from  {checkpoint}")
+        else: 
+            checkpoint = Path(checkpoint)
         if metric_type in checkpoint.name:
-            init_epoch = int(checkpoint.name.split(metric_type)[1][1:-3])
+            init_epoch = int(str(checkpoint.name).split(metric_type)[1].split(".keras")[0][1:])
         if (not reinitialise_momentum) & (metric_type in checkpoint.name):
+            print("loading model with states ")
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion
            print("loading model with optimizer states.")
```

File: SynthSeg/metrics_model.py [hunk 5]
```diff
@@ -113,7 +114,7 @@ def get_config(self):
         }
         return config
 
-    def __call__(self, gt, pred):
+    def __call__(self, gt, pred, **kwargs):
```
Comment (dcfidalgo) [inline SynthSeg/metrics_model.py:118]: Out of curiosity, do you know what these `kwargs` are?

File: SynthSeg/metrics_model.py [hunk 6]
```diff
@@ -99,8 +99,9 @@ def __call__(self, gt, pred):
         ) / (tf.keras.backend.sum(weights) * self._n_labels)
 
 
-class DiceLoss:
+class DiceLoss(tf.keras.losses.Loss):
```
Comment (dcfidalgo) [inline SynthSeg/metrics_model.py:103]: Could you do the same for the `WeightedL2Loss`, in case we decide to apply this loss for the training?

File: SynthSeg/training_with_tfrecords.py [hunk 7]
```diff
@@ -183,21 +228,87 @@ def load_model(
...
                 if key != "Layer"
             }
+
             custom_objects = {
                 **custom_l2i,
                 **custom_nrn,
                 "tf": tf,
                 "keras": tf.keras,
                 "loss": IdentityLoss().loss,
+                "DiceLoss": DiceLoss
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: What happens if we want to continue training for a model with the `WeightedL2Loss`? I guess we have add either the dice or the l2 loss, or maybe both, not sure.

File: SynthSeg/training_with_tfrecords.py [hunk 8]
```diff
@@ -206,16 +317,22 @@ def build_callbacks(
...
     # model saving callback
-    save_file_name = os.path.join(output_dir, "%s_{epoch:03d}.h5" % metric_type)
-    # save_weights_only=True is a workaround when training with mixed precision: https://github.com/keras-team/tf-keras/issues/203
-    # For now, setting this to true disables the posibility to train a model further, at least if we dont want to reinit the momentum ...
-    callbacks = [tf.keras.callbacks.ModelCheckpoint(save_file_name, verbose=1, save_weights_only=True)]
+    save_file_name = os.path.join(output_dir, "%s_{epoch:03d}.keras" % metric_type)
+
+
+    tracking_file = os.path.join(output_dir, f"time_recorder-num_gpu_{num_gpu}-strategy-{training_opts.strategy}_scaling-{training_opts.scaling_type}_steps_per_epoch-{steps_per_epoch}.csv")
+    tracking_callback = TimingCallback(tracking_file)
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: Again, i would prefer to leave this callback out of this PR.

File: SynthSeg/training_with_tfrecords.py [hunk 9]
```diff
@@ -66,8 +75,20 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
...
+
+    num_gpu  = strategy.num_replicas_in_sync if opts.strategy != "none" else len(tf.config.list_physical_devices('GPU'))
+    
+    if opts.scaling_type == "strong" and strategy is not None:
+            steps_per_epoch  = opts.steps_per_epoch//num_gpu
+            print(f"Number of steps per epoch was scaled to {opts.steps_per_epoch}")
+    elif opts.scaling_type ==" weak" and strategy is not None: 
+        steps_per_epoch = opts.steps_per_epoch
+    else: 
+        steps_per_epoch = opts.steps_per_epoch
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: See comment from above

File: SynthSeg/training_with_tfrecords.py [hunk 10]
```diff
@@ -30,6 +37,8 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
         opts: The training options. The parameters related to the generation of the synthetic images will be ignored.
     """
     # Check epochs
+    set_seed()
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: I would reuse this keras method which does the same: https://www.tensorflow.org/api_docs/python/tf/keras/utils/set_random_seed

File: SynthSeg/training_options.py [hunk 11]
```diff
@@ -368,6 +368,24 @@ class TrainingOptions(Serializable, OptionsBase):
...
+    find_last_checkpoint: bool = True
```
Comment (dcfidalgo) [inline SynthSeg/training_options.py]: I would prefer to set this as `False` to not change the current default behavior.

File: container/requirements_container.txt
Comment (dcfidalgo) [inline container/requirements_container.txt:1]: I would remove this file and add the requirements directly to the `.def` file.

**Review by dcfidalgo — COMMENTED** [review]
Nice, great work! I left a few inline comments. Regarding resuming the training with a WeightedL2Loss, i guess we will never really need it, since this loss was only used for some kind of "short pertaining" (actually i did not use it at all for our last training run). So either we implement it correctly, although we most likely wont use it, or we add a check in the training function, that throws an exception if one tries to resume a training with the wl2 loss. Your call :)

File: SynthSeg/metrics_model.py (hunk 5 above)
Comment (nhorlava) [reply SynthSeg/metrics_model.py:118]: It's "sample_weight" (https://github.com/keras-team/keras/blob/601488fd4c1468ae7872e132e0f1c9843df54182/keras/losses.py#L102C40-L102C53). I thought of using **kwargs for future version compatibility in case it changes in tf. Do you think we can keep it or change it to "sample_weight"?

File: SynthSeg/training_options.py (hunk 1 above)
Comment (nhorlava) [reply SynthSeg/training_options.py]: done

File: SynthSeg/training_with_tfrecords.py (hunk 2 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: removed:)

File: SynthSeg/training_with_tfrecords.py (hunk 3 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: great suggestion, done!

File: SynthSeg/training_with_tfrecords.py (hunk 8 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: done

File: SynthSeg/training_with_tfrecords.py (hunk 9 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: removed:)

File: SynthSeg/training_options.py (hunk 11 above)
Comment (nhorlava) [reply SynthSeg/training_options.py]: done

File: SynthSeg/training_with_tfrecords.py (hunk 10 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: ahh nice catch, replaced with keras method

File: container/requirements_container.txt
Comment (nhorlava) [reply container/requirements_container.txt:1]: done

File: SynthSeg/metrics_model.py (hunk 6 above)
Comment (nhorlava) [reply SynthSeg/metrics_model.py:103]: done

**Comment by nhorlava** [comment]
To address this comment: 
>  Regarding resuming the training with a WeightedL2Loss, i guess we will never really need it, since this loss was only used for some kind of "short pertaining" (actually i did not use it at all for our last training run). So either we implement it correctly, although we most likely won't use it, or we add a check in the training function, that throws an exception if one tries to resume a training with the wl2 loss. 

I went for the "check" option. So if the checkpoint is provided in training options, the code expects it to be a checkpoint from training with dice loss. That is checked by checking that "dice" metric is in the checkpoint name, and would throw an assertion error otherwise:
https://github.com/halirutan/SynthSeg/blob/ff249662fd509efad704d0e12507e77792cd1e05/SynthSeg/training_with_tfrecords.py#L224
**Note**: in case `find_last_checkpoint==True`, only checkpoints from training with Dice loss would be considered when looking for the last available checkpoint:
https://github.com/halirutan/SynthSeg/blob/ff249662fd509efad704d0e12507e77792cd1e05/SynthSeg/training_with_tfrecords.py#L204

If in your provided training options wl2_epochs>0 and checkpoint from dice training provided, the code will assume that pre-training with wl2 loss already happened, as well as N epochs of training with dice loss. Therefore, pre-training with WL2 loss will be skipped, and training will resume from checkpoint and N+1 epoch with Dice loss. :
https://github.com/halirutan/SynthSeg/blob/ff249662fd509efad704d0e12507e77792cd1e05/SynthSeg/training_with_tfrecords.py#L124
_____

See tests that resuming works correctly
1) when only training with Dice Loss
image

2) When pre-training with WL2 loss, and continuing with Dice Loss
image

File: SynthSeg/metrics_model.py (hunk 5 above)
Comment (dcfidalgo) [reply SynthSeg/metrics_model.py:118]: kwargs is perfect, I was just curious :)

File: SynthSeg/training_options.py [hunk 12]
```diff
@@ -368,6 +368,20 @@ class TrainingOptions(Serializable, OptionsBase):
     See https://www.tensorflow.org/guide/distributed_training for more information. 
     """
 
+    find_last_checkpoint: bool = False
+    """
+    If set to True, would assume that 'checkpoint' is a directory and would try to find the last available checkpoint
+    by skanning all files with extension ".keras", getting the epoch by the name.
+    In case of multiple filenames from the same epoch the file that was created teh last will be chosen.
+
+    If set to False, 'checkpoint' parameter will be treated like a pull path to the checkpoint file.
```
Comment (dcfidalgo) [inline SynthSeg/training_options.py]:
```suggestion
    If set to True, assume that 'checkpoint' is a directory and try to find the last available checkpoint
    by scanning all the files with the extension ".keras", getting the epoch from the name.
    In case of multiple filenames from the same epoch, the last file created will be chosen.

    If set to False, 'checkpoint' parameter will be treated like a full path to the checkpoint file. 
```

File: SynthSeg/training_options.py [hunk 13]
```diff
@@ -400,5 +414,7 @@ def get_non_path_string_options() -> List[str]:
             "prior_distributions",
             "wandb_log_freq",
             "compression_type",
-            "strategy"
+            "strategy", 
+            "find_last_checkpoint", 
+            "seed"
```
Comment (dcfidalgo) [inline SynthSeg/training_options.py]:
```suggestion
            "strategy"
```
No need to add those, since they are not strings

File: SynthSeg/training_with_tfrecords.py [hunk 14]
```diff
@@ -122,25 +139,36 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
...
+                load_after_wl2_pretraining = load_after_wl2_pretraining
             )
+
             if not is_compiled:
                 dice_model.compile(
                     optimizer=tf.keras.optimizers.Adam(learning_rate=opts.lr),
                     loss=DiceLoss(n_labels=opts.n_labels),
                 )
 
+        print(f"Optimizers  number of iterations at the beginning: ",  dice_model.optimizer.iterations.numpy())
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion
        print(f"Optimizers number of iterations at the beginning: ",  dice_model.optimizer.iterations.numpy())
```

File: SynthSeg/training_with_tfrecords.py [hunk 15]
```diff
@@ -122,25 +139,36 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
...
+        print(f"Amount of batches that will be skipped: {opts.wl2_epochs*opts.steps_per_epoch} batches from wl2 pretraining + {init_batch} from training on dice loss")
+        print(f"Restarting from heckpoint: {init_epoch}")
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion
        print(f"Restarting from checkpoint: {init_epoch}")
```

File: SynthSeg/training_with_tfrecords.py [hunk 16]
```diff
@@ -99,13 +110,19 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
...
+            
             wl2_model.compile(
                 optimizer=tf.keras.optimizers.Adam(learning_rate=opts.lr),
                 loss=WeightedL2Loss(n_labels=opts.n_labels),
             )
+            
 
     results = None
-    if opts.wl2_epochs > 0:
+    # additionally check that chackpoint is None, otherwise we assume that we want to
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion
    # additionally check that checkpoint is None, otherwise we assume that we want to
```

**Review by dcfidalgo — APPROVED** [review]
Very nice! :100: Just a few suggestions that I will apply myself. 
Will merge this now
